go.mod

```diff
@@ -5,10 +5,11 @@ go 1.13
 require (
 	github.com/dustinkirkland/golang-petname v0.0.0-20190613200456-11339a705ed2
 	github.com/go-logr/logr v0.1.0
+	github.com/google/go-cmp v0.3.0
 	github.com/stretchr/testify v1.3.0
-	k8s.io/api v0.0.0-20190409021203-6e4e0e4f393b
-	k8s.io/apimachinery v0.0.0-20190404173353-6a84e37a896d
-	k8s.io/client-go v11.0.1-0.20190409021438-1a26190bd76a+incompatible
-	sigs.k8s.io/controller-runtime v0.2.2
-	sigs.k8s.io/controller-tools v0.2.1 // indirect
+	k8s.io/api v0.0.0-20190918195907-bd6ac527cfd2
+	k8s.io/apimachinery v0.0.0-20190817020851-f2f3a405f61d
+	k8s.io/client-go v0.0.0-20190918200256-06eb1244587a
+	k8s.io/utils v0.0.0-20190506122338-8fab8cb257d5
+	sigs.k8s.io/controller-runtime v0.3.0
```

Comment:
A quick explanation.
* CheckStructFields was difficult to test, because it contained calls to `assert.NoError` and `assert.Equal`,
* So I refactored it to return an error instead.
* I introduced `go-cmp`  as an alternative to testify `assert.Equal` and also because it handles pointers better than assert.Equal.
* `go-cmp` was then failing to compare `resource.Quantity` (something to do with it having pivate struct fields)
* Newer versions of apimachinery, have added a `go-cmp` compatible `Equals` method to `resource.Quantity`, which fixes with the last problem.
* I struggled to find an apimachinery which was compatible with all the other `k8s` dependencies, 
* So in the end decided to update controller-runtime to `0.3.0` (latest stable) and to use all the same versions of `k8s.io` dependencies that it uses, as advised in https://github.com/kubernetes-sigs/controller-runtime/issues/633#issuecomment-540681752

So these are the versions from: https://github.com/kubernetes-sigs/controller-runtime/blob/v0.3.0/go.mod

Reply:
what a rabbit hole 😁 out of curiosity, did you experiment with `reflect.DeepEqual` ?

Reply:
DeepEqual doesn't handle resource.Quantity, I think.
Because two quantities might have different SI units, but have the same value.